File: sdks/code-interpreter/kotlin/code-interpreter/src/main/kotlin/com/alibaba/opensandbox/codeinterpreter/infrastructure/adapters/service/CodesAdapter.kt
/*
 * Copyright 2025 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.codeinterpreter.infrastructure.adapters.service

import com.alibaba.opensandbox.codeinterpreter.domain.models.execd.executions.CodeContext
import com.alibaba.opensandbox.codeinterpreter.domain.models.execd.executions.RunCodeRequest
import com.alibaba.opensandbox.codeinterpreter.domain.services.Codes
import com.alibaba.opensandbox.codeinterpreter.infrastructure.adapters.converter.CodeExecutionConverter.toApiRunCodeRequest
import com.alibaba.opensandbox.codeinterpreter.infrastructure.adapters.converter.CodeExecutionConverter.toCodeContext
import com.alibaba.opensandbox.sandbox.HttpClientProvider
import com.alibaba.opensandbox.sandbox.api.execd.CodeInterpretingApi
import com.alibaba.opensandbox.sandbox.api.models.execd.EventNode
import com.alibaba.opensandbox.sandbox.domain.exceptions.InvalidArgumentException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxApiException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxError
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxError.Companion.UNEXPECTED_RESPONSE
import com.alibaba.opensandbox.sandbox.domain.models.execd.executions.Execution
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxEndpoint
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.converter.ExecutionEventDispatcher
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.converter.jsonParser
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.converter.parseSandboxError
import com.alibaba.opensandbox.sandbox.infrastructure.adapters.converter.toSandboxException
import okhttp3.Headers.Companion.toHeaders
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.slf4j.LoggerFactory
import com.alibaba.opensandbox.sandbox.api.models.execd.CodeContextRequest as ApiCodeContextRequest

class CodesAdapter(
    private val execdEndpoint: SandboxEndpoint,
    private val httpClientProvider: HttpClientProvider,
) : Codes {
    companion object {
        private const val RUN_CODE_PATH = "/code"
    }

    private val logger = LoggerFactory.getLogger(CodesAdapter::class.java)
    private val baseUrl = "${httpClientProvider.config.protocol}://${execdEndpoint.endpoint}"
    private val apiClient =
        httpClientProvider.httpClient.newBuilder()
            .addInterceptor { chain ->
                val requestBuilder = chain.request().newBuilder()
                execdEndpoint.headers.forEach { (key, value) ->
                    requestBuilder.header(key, value)
                }
                chain.proceed(requestBuilder.build())
            }
            .build()
    private val api =
        CodeInterpretingApi(baseUrl, apiClient)

    override fun getContext(id: String): CodeContext {
        try {
            val result = api.getContext(id)
            return result.toCodeContext()
        } catch (e: Exception) {
            logger.error("Failed to get context", e)
            throw e.toSandboxException()
        }
    }

    override fun listContexts(language: String): List<CodeContext> {
        try {
            val list = api.listContexts(language)
            return list.map { it.toCodeContext() }
        } catch (e: Exception) {
            logger.error("Failed to list contexts", e)
            throw e.toSandboxException()
        }
    }

    override fun createContext(language: String): CodeContext {
        try {
            val request = ApiCodeContextRequest(language = language)
            val result = api.createCodeContext(request)
            return result.toCodeContext()
        } catch (e: Exception) {
            logger.error("Failed to create context", e)
            throw e.toSandboxException()
        }
    }

    override fun deleteContext(id: String) {
        try {
            api.deleteContext(id)
        } catch (e: Exception) {
            logger.error("Failed to delete context", e)
            throw e.toSandboxException()
        }
    }

    override fun deleteContexts(language: String) {
        try {
            deleteContexts(language)
        } catch (e: Exception) {
            logger.error("Failed to delete contexts", e)
            throw e.toSandboxException()
        }
    }

    override fun run(request: RunCodeRequest): Execution {
        if (request.code.isEmpty()) {
            throw InvalidArgumentException("Code cannot be empty")
        }
        try {
            val apiRequest = request.toApiRunCodeRequest()
            val httpRequest =
                Request.Builder()
                    .url("$baseUrl$RUN_CODE_PATH")
                    .post(
                        jsonParser.encodeToString(apiRequest).toRequestBody("application/json".toMediaType()),
                    )
                    .headers(execdEndpoint.headers.toHeaders())
                    .build()

            val execution = Execution()

            httpClientProvider.sseClient.newCall(httpRequest).execute().use { response ->
                if (!response.isSuccessful) {
                    val errorBodyString = response.body?.string()
                    val sandboxError = parseSandboxError(errorBodyString)
                    val message = "Failed to run code. Status code: ${response.code}, Body: $errorBodyString"
                    throw SandboxApiException(
                        message = message,
                        statusCode = response.code,
                        error = sandboxError ?: SandboxError(UNEXPECTED_RESPONSE),
                        requestId = response.header("X-Request-ID"),
                    )
                }

                response.body?.byteStream()?.bufferedReader(Charsets.UTF_8)?.use { reader ->
                    val dispatcher = ExecutionEventDispatcher(execution, request.handlers)
                    reader.lineSequence()
                        .filter(String::isNotBlank)
                        .forEach { line ->
                            try {
                                val eventNode = jsonParser.decodeFromString<EventNode>(line)
                                dispatcher.dispatch(eventNode)
                            } catch (e: Exception) {
                                logger.error("Failed to parse SSE line: {}", line, e)
                            }
                        }
                }
            }

            return execution
        } catch (e: Exception) {
            logger.error("Failed to run code (length: {})", request.code.length, e)
            throw e.toSandboxException()
        }
    }

    override fun interrupt(executionId: String) {
        try {
            api.interruptCode(executionId)
        } catch (e: Exception) {
            logger.error("Failed to interrupt code execution", e)
            throw e.toSandboxException()
        }
    }
}

File: sdks/code-interpreter/csharp/src/OpenSandbox.CodeInterpreter/Adapters/CodesAdapter.cs
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;
using OpenSandbox.Adapters;
using OpenSandbox.CodeInterpreter.Models;
using OpenSandbox.CodeInterpreter.Services;
using OpenSandbox.Core;
using OpenSandbox.Internal;
using OpenSandbox.Models;
using Microsoft.Extensions.Logging;

namespace OpenSandbox.CodeInterpreter.Adapters;

/// <summary>
/// Adapter implementation for the codes service.
/// </summary>
internal sealed class CodesAdapter : ICodes
{
    private readonly HttpClientWrapper _client;
    private readonly HttpClient _sseHttpClient;
    private readonly string _baseUrl;
    private readonly IReadOnlyDictionary<string, string> _headers;
    private readonly ILogger _logger;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
    };

    public CodesAdapter(
        HttpClientWrapper client,
        HttpClient sseHttpClient,
        string baseUrl,
        IReadOnlyDictionary<string, string> headers,
        ILogger logger)
    {
        _client = client ?? throw new ArgumentNullException(nameof(client));
        _sseHttpClient = sseHttpClient ?? throw new ArgumentNullException(nameof(sseHttpClient));
        _baseUrl = baseUrl?.TrimEnd('/') ?? throw new ArgumentNullException(nameof(baseUrl));
        _headers = headers ?? new Dictionary<string, string>();
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    public async Task<CodeContext> CreateContextAsync(string language, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(language))
        {
            throw new InvalidArgumentException("Language cannot be empty");
        }

        var request = new CreateContextRequest { Language = language };
        _logger.LogDebug("Creating code context (language={Language})", language);
        var response = await _client.PostAsync<CodeContext>("/code/context", request, cancellationToken).ConfigureAwait(false);

        if (response == null || string.IsNullOrEmpty(response.Language))
        {
            throw new SandboxApiException(
                message: "Create code context failed: unexpected response shape",
                error: new SandboxError(SandboxErrorCodes.UnexpectedResponse, "Create code context failed: unexpected response shape"));
        }

        return response;
    }

    public async Task<CodeContext> GetContextAsync(string contextId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(contextId))
        {
            throw new InvalidArgumentException("contextId cannot be empty");
        }

        _logger.LogDebug("Fetching code context: {ContextId}", contextId);
        var response = await _client.GetAsync<CodeContext>($"/code/contexts/{Uri.EscapeDataString(contextId)}", cancellationToken: cancellationToken).ConfigureAwait(false);

        if (response == null || string.IsNullOrEmpty(response.Language))
        {
            throw new SandboxApiException(
                message: "Get code context failed: unexpected response shape",
                error: new SandboxError(SandboxErrorCodes.UnexpectedResponse, "Get code context failed: unexpected response shape"));
        }

        return response;
    }

    public async Task<IReadOnlyList<CodeContext>> ListContextsAsync(string language, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(language))
        {
            throw new InvalidArgumentException("Language cannot be empty");
        }

        _logger.LogDebug("Listing code contexts (language={Language})", language);
        var queryParams = new Dictionary<string, string?> { ["language"] = language };

        var response = await _client.GetAsync<List<CodeContext>>("/code/contexts", queryParams, cancellationToken).ConfigureAwait(false);

        if (response == null)
        {
            throw new SandboxApiException(
                message: "List code contexts failed: unexpected response shape",
                error: new SandboxError(SandboxErrorCodes.UnexpectedResponse, "List code contexts failed: unexpected response shape"));
        }

        return response;
    }

    public async Task DeleteContextAsync(string contextId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(contextId))
        {
            throw new InvalidArgumentException("contextId cannot be empty");
        }

        _logger.LogInformation("Deleting code context: {ContextId}", contextId);
        await _client.DeleteAsync($"/code/contexts/{Uri.EscapeDataString(contextId)}", cancellationToken: cancellationToken).ConfigureAwait(false);
    }

    public async Task DeleteContextsAsync(string language, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(language))
        {
            throw new InvalidArgumentException("Language cannot be empty");
        }

        _logger.LogInformation("Deleting code contexts (language={Language})", language);
        var queryParams = new Dictionary<string, string?> { ["language"] = language };
        await _client.DeleteAsync("/code/contexts", queryParams, cancellationToken).ConfigureAwait(false);
    }

    public async Task InterruptAsync(string executionId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(executionId))
        {
            throw new InvalidArgumentException("executionId cannot be empty");
        }

        _logger.LogInformation("Interrupting code execution: {ExecutionId}", executionId);
        var queryParams = new Dictionary<string, string?> { ["id"] = executionId };
        await _client.DeleteAsync("/code", queryParams, cancellationToken).ConfigureAwait(false);
    }

    public async IAsyncEnumerable<ServerStreamEvent> RunStreamAsync(
        RunCodeRequest request,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        if (request == null)
        {
            throw new InvalidArgumentException("request cannot be null");
        }

        if (string.IsNullOrWhiteSpace(request.Code))
        {
            throw new InvalidArgumentException("Code cannot be empty");
        }

        var url = $"{_baseUrl}/code";
        _logger.LogDebug("Running code stream (codeLength={CodeLength})", request.Code.Length);
        var json = JsonSerializer.Serialize(request, JsonOptions);

        using var httpRequest = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new StringContent(json, Encoding.UTF8, "application/json")
        };

        httpRequest.Headers.Accept.Add(new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("text/event-stream"));

        foreach (var header in _headers)
        {
            httpRequest.Headers.TryAddWithoutValidation(header.Key, header.Value);
        }

        using var response = await _sseHttpClient.SendAsync(httpRequest, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);

        await foreach (var ev in SseParser.ParseJsonEventStreamAsync<ServerStreamEvent>(response, "Run code failed", cancellationToken).ConfigureAwait(false))
        {
            yield return ev;
        }
    }

    public async Task<Execution> RunAsync(string code, RunCodeOptions? options = null, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(code))
        {
            throw new InvalidArgumentException("Code cannot be empty");
        }

        if (options?.Context != null && options.Language != null)
        {
            throw new InvalidArgumentException("Provide either options.Context or options.Language, not both");
        }

        var context = options?.Context
            ?? (options?.Language != null
                ? new CodeContext { Language = options.Language }
                : new CodeContext { Language = SupportedLanguage.Python });

        var request = new RunCodeRequest
        {
            Code = code,
            Context = context
        };

        var execution = new Execution();
        _logger.LogDebug("Running code (codeLength={CodeLength})", code.Length);
        var dispatcher = new ExecutionEventDispatcher(execution, options?.Handlers);

        await foreach (var ev in RunStreamAsync(request, cancellationToken).ConfigureAwait(false))
        {
            await dispatcher.DispatchAsync(ev).ConfigureAwait(false);
        }

        return execution;
    }
}
